SIPS/sip-77.md

```diff
@@ -0,0 +1,165 @@
+- `notifyRewardAmount`
+
+The reward rate is set inside `notifyRewardAmount`, if a value that is too large is provided to the function.
+Of particular note is that `notifyRewardAmount` is itself affected by this problem, which means that if the provided
+reward is incorrect, then the problem is unrecoverable.
+
+#### Solution
+
+The `notifyRewardAmount` transaction should be reverted if a value greater than `2^256 / 10^18` is provided.
+As an additional safety mechanism, this value will be required to be no greater than the remaining
+balance of the rewards token in the contract. This will both prevent the overflow, and also provide an additional check
+that the reward rate is being set to a value in the appropriate range (for example, no extra/missing zeroes).
+
+#### Details
+
+Specifically, this problem occurs when rewardRate is too high; it is set inside the `notifyRewardAmount` function on
+lines [114](https://github.com/Synthetixio/synthetix/blob/c4dd4413cbbd3c0b40dfee2f9119af2dcb6a82e5/contracts/StakingRewards.sol#L114) and [118](https://github.com/Synthetixio/synthetix/blob/c4dd4413cbbd3c0b40dfee2f9119af2dcb6a82e5/contracts/StakingRewards.sol#L118).
+
+```
+rewardRate = floor(reward / rewardsDuration) = (reward - k) / rewardsDuration
+```
+
+for some `0 <= k < rewardsDuration`.
+
+For the bug to occur, we need:
+
+```
+(reward - k) / rewardsDuration >= 2^256 / (10^18 * (lastTimeRewardApplicable - lastUpdateTime))
+reward                         >= rewardsDuration * 2^256 / (10^18 * (lastTimeRewardApplicable - lastUpdateTime)) + k
+```
+
+Hence, we can ensure the bug does not occur if we force:
+
+```
+reward < rewardsDuration * 2^256 / (10^18 * (lastTimeRewardApplicable - lastUpdateTime))
+```
+
+So we should constrain `reward` to be less than the minimum value of the RHS.
+
+The smallest possible value of `lastUpdateTime` is the block timestamp when `notifyRewardAmount` was last called.
+The largest possible value of `lastTimeRewardApplicable` is `periodFinish`,
+and `periodFinish = notificationBlock.timestamp + rewardsDuration` ([line 121](https://github.com/Synthetixio/synthetix/blob/c4dd4413cbbd3c0b40dfee2f9119af2dcb6a82e5/contracts/StakingRewards.sol#L121)).
+Putting these together we have:
+
+```
+(lastTimeRewardApplicable - lastUpdateTime) <= rewardsDuration
+```
+
+Ergo, we need:
+
+```
+reward < rewardsDuration * 2^256 / (10^18 * rewardsDuration)
+	                     = 2^256 / 10^18
+```
+
+So the problem will not emerge whenever we require
+
+```
+    reward < uint(-1) / UNIT
+```
+
+### Fix to setRewardsDuration to allow updates after the initial setting
+
+`setRewardsDuration` was intended to allow the `rewardsDuration` to be set after theduration had completed. However a flaw in the require meant it could be changed after the initial setting.
+
+Current code
+
+```
+require(periodFinish == 0 || block.timestamp > periodFinish);
+```
+
+Proposed change
+
+```
+require(periodFinish == 0 && block.timestamp > periodFinish);
```

Comment:
Can remove the `periodFinish == 0` check and only check that block.timestamp is after periodFinish